Vulkan Game Capture

LittleMatth opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
add support for vulkan game capture.
graphics-hook dll now implement a vulkan implicit layer which gets loaded by vulkan runtime if registered properly.
this allow obs to access a shared texture containing game's output
<!--- If this change includes UI elements, please include screenshots. -->

### Motivation and Context
capturing vulkan games
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open Mantis issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->

### How Has This Been Tested?
tested with multiple commercial vulkan games ( doom, no man sky, dota2, wolfenstein)
 on both nvidia (1070, 2070) and amd adapters (R9 290, RX5700).
In order to capture vulkan games you need to have the layer registered in windows registry
you may use the batch file located here : data\obs-plugins\win-capture\obs-vulkan-registry.bat

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
 - New feature (non-breaking change which adds functionality)
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [ ] My code is not on the master branch.
- [x] The code has been tested.
- [ ] All commit messages are properly formatted and commits squashed where appropriate.

LittleMatth commented:
things to do before pulling 
 - implement a programatic, transparent way to register/unregister the vulkan layer
 - test more vulkan games on more amd gpu, and optionnally update the format selection function.

Sunderland93 commented:
Linux support?

LittleMatth commented:
> ### Description
> add support for vulkan game capture.
> graphics-hook dll now implement a vulkan implicit layer which gets loaded by vulkan runtime if registered properly.
> this allow obs to access a shared texture containing game's output
> 
> ### Motivation and Context
> capturing vulkan games
> 
> ### How Has This Been Tested?
> tested with multiple commercial vulkan games ( doom, no man sky, dota2, wolfenstein)
> on both nvidia (1070, 2070) and amd adapters (R9 290, RX5700).
> In order to capture vulkan games you need to have the layer registered in windows registry
> you may use the batch file located here : data\obs-plugins\win-capture\obs-vulkan-registry.bat
> 
> ### Types of changes
> * New feature (non-breaking change which adds functionality)
> 
> ### Checklist:
> * [x]  My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
> * [x]  I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
> * [ ]  My code is not on the master branch.
> * [x]  The code has been tested.
> * [ ]  All commit messages are properly formatted and commits squashed where appropriate.

LittleMatth commented:
> Linux support?
i didn't spend time investigating other platform than windows.
but based on what i did, it should help adding the linux support.

jp9000 commented:
This is a windows-only implementation at the moment.  Open source is about taking what you're able to get sometimes, and Linux already is lacking a game-hooking framework as it is right now.  I don't like it when Linux/macOS aren't able to get the same treatment as Windows, but sometimes we have to just take what we can get.  In the future perhaps we'll have more capture options for Linux (and perhaps we might bounty out a game capturing framework in general), but for now it's just Windows.

jp9000 commented:
Oh yea that's right, Vulkan uses a layer framework, not hooking.  The layering framework should make it much easier to implement a vulkan capture on Linux in the future as well.  So while we won't have it for quite some time, it'll make it easier to implement in the future.

Sunderland93 commented:
@jp9000 

> game-hooking framework

Can you please explain what is it?

jp9000 commented:
@Sunderland93 
> Can you please explain what is it?

I apologize, it doesn't require a hooking framework as Vulkan uses its own framework called layers to attach extra render code to games.  However, we still don't have any general code for game capturing on Linux, and I am not going to impose upon this author that he do all that work in addition, nor do I want anyone else to do so either.

That being said, from what it seems it will be possible to hook Linux games in the future without that much difficulty thanks to layering, but we are not going to do it in this pull request.  The module this modifies is windows-only anyway, so it would not be possible to implement a Linux version right now or on this pull request.  Later on, if there is a Linux developer who is skilled an willing, they will be able to learn from this code and make a similar capture on Linux without too much difficulty.  But this will not happen right now or on this pull request.

topeterk commented:
Is there a way of getting a pre-compiled test version of the plugin already?
Want to try capturing The Surge 2 running on i7-4771 with GeForce 1060 GTX.

Q: For the OBS-Studio maintainers: Kindly ask for estimate when such a feature may be integrated into an official version? (Days, Weeks, Months?)

topeterk commented:
Couldn't resist: VM installed with Win10, downloaded all sources, IDEs and more dependencies, patched commits of this issue and compiled it.
After running the batch, starting OBS Studio and then starting the game, I got a video from the game capture of "_The Surge 2 (Vulkan, 64Bit)_". This is great!

If someone else wants to try this, I attached my built **Win10x64** binaries of the folder "_data\obs-plugins\win-capture_" which should be enough to get it running. Remember to run the batch to activate the hook (and start OBS before the game).
[OBSStudio24.0.3 VulkanGameGapture-caff574 Win10x64.zip](https://github.com/obsproject/obs-studio/files/3870902/OBSStudio24.0.3.VulkanGameGapture-caff574.Win10x64.zip)

Opt1kal commented:
> Couldn't resist: VM installed with Win10, downloaded all sources, IDEs and more dependencies, patched commits of this issue and compiled it.
> After running the batch, starting OBS Studio and then starting the game, I got a video from the game capture of "_The Surge 2 (Vulkan, 64Bit)_". This is great!
> 
> If someone else wants to try this, I attached my built **Win10x64** binaries of the folder "_data\obs-plugins\win-capture_" which should be enough to get it running. Remember to run the batch to activate the hook (and start OBS before the game).
> [OBSStudio24.0.3 VulkanGameGapture-caff574 Win10x64.zip](https://github.com/obsproject/obs-studio/files/3870902/OBSStudio24.0.3.VulkanGameGapture-caff574.Win10x64.zip)

Just wondering, what would I need to do to convert this over to picking up Red Dead Redemption 2's Vulkan? 

Looking at the capture drop down in OBS-Studio, it did list The Surge 2(Vulkan) as an option originally, but when selecting Red Dead Redemption 2, it's still showing the black screen. Tried looking in the json's and in the batch file, couldn't see anything that stood out to me as being specific to The Surge 2.

ItsiAdam commented:
rdr2 has an odd vulkan implementation, someone more experienced with vk needs to have a look at this

LittleMatth commented:
Any game developper may prevent the loading of unidentified implicit vulkan layer. it is completely possible that rockstar prohibit this at the moment.

jp9000 commented:
I will help on this PR soon.  I just need to work on other pull requests a bit more.

topeterk commented:
Hi, don't know if this is a thing but I noticed that the steam overlay (friend comes online/achievement) as well as the NVIDIA FPS counter from Geforce Experience are flickering in the recording or from time to time disappear at all or is fully present.
Don't know much about Vulkan but I thought this is some kind of render pipeline you can integrate yourself into but this would mean the "hook" is running always at the same time. Or is this kind of random due to parallelism, so the GPU desides when to call which hook "in the line"?
Just wanted to let you know about this - maybe this needs attention or not (as by design). However, when it could be improved that the capturing is kind of "stable", this would be great for OBS use.

LittleMatth commented:
> Hi, don't know if this is a thing but I noticed that the steam overlay (friend comes online/achievement) as well as the NVIDIA FPS counter from Geforce Experience are flickering in the recording or from time to time disappear at all or is fully present.
> Don't know much about Vulkan but I thought this is some kind of render pipeline you can integrate yourself into but this would mean the "hook" is running always at the same time. Or is this kind of random due to parallelism, so the GPU desides when to call which hook "in the line"?
> Just wanted to let you know about this - maybe this needs attention or not (as by design). However, when it could be improved that the capturing is kind of "stable", this would be great for OBS use.

i did not spend time investigating how to control the layer's order, but as far as i understood, sequencing of the different layers should be stable during a single run and the order should be controlled by the vulkan loader.

Leopard1907 commented:
If what people posted on various forums are true , RDR2 uses a d3d11 interop on both D3D12 and VLK backends.

radum706 commented:
I just tested it and i still have drop frames

radum706 commented:
> I just tested it and i still have drop frames

I forgot to mention that I tested it in No Man's Sky

Anths2016 commented:
Retroarch crash if I change game.

Fenrirthviti commented:
A build can be found here for testing: https://dev.azure.com/obsjim/obsjim/_build/results?buildId=2156

Click on Artifacts, and then Winbuild to download.

Mister36MN commented:
Will get this build and test within the next couple of days

Mister36MN commented:
Got it working after some finagling with the batch, capture worked well. (Doom 2016)

Anths2016 commented:
> Got it working after some finagling with the batch, capture worked well. (Doom 2016)

Emulators crash if you close them.

AluminumHaste commented:
> 
> 
> Couldn't resist: VM installed with Win10, downloaded all sources, IDEs and more dependencies, patched commits of this issue and compiled it.
> After running the batch, starting OBS Studio and then starting the game, I got a video from the game capture of "_The Surge 2 (Vulkan, 64Bit)_". This is great!
> 
> If someone else wants to try this, I attached my built **Win10x64** binaries of the folder "_data\obs-plugins\win-capture_" which should be enough to get it running. Remember to run the batch to activate the hook (and start OBS before the game).
> [OBSStudio24.0.3 VulkanGameGapture-caff574 Win10x64.zip](https://github.com/obsproject/obs-studio/files/3870902/OBSStudio24.0.3.VulkanGameGapture-caff574.Win10x64.zip)

This works great for 32bit games, but breaks Vulkan 64bit apps.
Including 64bit version of cube.exe. 
I tried the 64bit and 32bit version of VKQuake (https://github.com/Novum/vkQuake) and once I register the json with the batch file, the 64bit versions of vulkan apps crash. The window appears, it's white, then after about 1-2 seconds, disappears.
32bit versions work fine and I can record them with OBS, works really well so far, so absolutely thank you for that.

My system is 
Windows 10 Pro v1909 running on Mushkin 1TB SSD.
AMD Ryzen 2700X @ whatever PBO fucking decides to run at
AMD Gigabyte RX 5700XT Gaming OC Latest drivers (19.12.3)

Would love to help get this working, please let me know what you need.

AluminumHaste commented:
I checked event viewer, not sure if this helps:

Faulting application name: vkcube.exe, version: 0.0.0.0, time stamp: 0x5dfd005c
Faulting module name: ntdll.dll, version: 10.0.18362.418, time stamp: 0x99ca0526
Exception code: 0xc0000409
Fault offset: 0x000000000008c4df
Faulting process id: 0xbe4
Faulting application start time: 0x01d5bf6f59a7e870
Faulting application path: C:\VulkanSDK\1.1.130.0\Bin\vkcube.exe
Faulting module path: C:\WINDOWS\SYSTEM32\ntdll.dll
Report Id: 47a50205-2c28-412c-bd6f-923dcca7a0b5
Faulting package full name: 
Faulting package-relative application ID:

LittleMatth commented:
> > Couldn't resist: VM installed with Win10, downloaded all sources, IDEs and more dependencies, patched commits of this issue and compiled it.
> > After running the batch, starting OBS Studio and then starting the game, I got a video from the game capture of "_The Surge 2 (Vulkan, 64Bit)_". This is great!
> > If someone else wants to try this, I attached my built **Win10x64** binaries of the folder "_data\obs-plugins\win-capture_" which should be enough to get it running. Remember to run the batch to activate the hook (and start OBS before the game).
> > [OBSStudio24.0.3 VulkanGameGapture-caff574 Win10x64.zip](https://github.com/obsproject/obs-studio/files/3870902/OBSStudio24.0.3.VulkanGameGapture-caff574.Win10x64.zip)
> 
> This works great for 32bit games, but breaks Vulkan 64bit apps.
> Including 64bit version of cube.exe.
> I tried the 64bit and 32bit version of VKQuake (https://github.com/Novum/vkQuake) and once I register the json with the batch file, the 64bit versions of vulkan apps crash. The window appears, it's white, then after about 1-2 seconds, disappears.
> 32bit versions work fine and I can record them with OBS, works really well so far, so absolutely thank you for that.
> 
> My system is
> Windows 10 Pro v1909 running on Mushkin 1TB SSD.
> AMD Ryzen 2700X @ whatever PBO fucking decides to run at
> AMD Gigabyte RX 5700XT Gaming OC Latest drivers (19.12.3)
> 
> Would love to help get this working, please let me know what you need.

Thoses feedbacks are really appreciated !
It seems strange to me that 64bit version of cube.exe fails, because
it was the test application i used to test the code i wrote.

If you are able to debug the running game (lauching the game or cube.exe with visual studio)
you should be able to break into vulkan-capture.c to understand what is going on.
But 32bit / 64bit needs different dll build + a specific vulkan json per target (one for 32 and one for 64)
maybe your 32bit dll is registered and gets loaded for 64bit application, this may cause issue.

AluminumHaste commented:
That's probably exactly it, I had issues with Vulkan being stupid before with Nvidia drivers breaking things.
I'll take a look at the registry when I get home from work.

alinsavix commented:
FWIW, I tested a self-built version of this with Detroit: Become Human (which is Vulkan-only). As long as I have OBS running before the game is running, it captures great!   ...until I change scenes to a scene that doesn't have my game capture source. Once that happens, I can't get the capture back until I restart the game, at which point the same pattern will repeat.

The log doesn't have much useful, just:
```
13:20:01.287: [game-capture: 'Game Capture'] capture stopped
```
I'm happy to gather whatever debugging information is needed if you'll let me know what you need.

AluminumHaste commented:
Okay I debug started with VS2017, here's the stack trace:

>	ntdll.dll!LdrpICallHandler()	Unknown
 	ntdll.dll!RtlpExecuteHandlerForException()	Unknown
 	ntdll.dll!RtlDispatchException()	Unknown
 	ntdll.dll!KiUserExceptionDispatch()	Unknown
 	ntdll.dll!LdrpDispatchUserCallTarget()	Unknown
 	vulkan-1.dll!00007fff4c00dbf0()	Unknown
 	graphics-hook64.dll!00007fff6a50b6a7()	Unknown
 	bdcamvk64.dll!00007fff6e593343()	Unknown
 	vulkan-1.dll!00007fff4c01afb7()	Unknown
 	vulkan-1.dll!00007fff4c0217f6()	Unknown
 	vulkan-1.dll!00007fff4c02d23d()	Unknown
 	vkcube.exe!00007ff6cc047e46()	Unknown
 	vkcube.exe!00007ff6cc049fd4()	Unknown
 	vkcube.exe!00007ff6cc046acf()	Unknown
 	vkcube.exe!00007ff6cc04f17e()	Unknown
 	kernel32.dll!00007fff933a7bd4()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown

I'm still trying to figure out the other things though

LittleMatth commented:
> Okay I debug started with VS2017, here's the stack trace:
> 
> > ntdll.dll!LdrpICallHandler�()	Unknown
> > ntdll.dll!RtlpExecuteHandlerForException�()	Unknown
> > ntdll.dll!RtlDispatchException()	Unknown
> > ntdll.dll!KiUserExceptionDispatch�()	Unknown
> > ntdll.dll!LdrpDispatchUserCallTarget�()	Unknown
> > vulkan-1.dll!00007fff4c00dbf0()	Unknown
> > graphics-hook64.dll!00007fff6a50b6a7()	Unknown
> > bdcamvk64.dll!00007fff6e593343()	Unknown
> > vulkan-1.dll!00007fff4c01afb7()	Unknown
> > vulkan-1.dll!00007fff4c0217f6()	Unknown
> > vulkan-1.dll!00007fff4c02d23d()	Unknown
> > vkcube.exe!00007ff6cc047e46()	Unknown
> > vkcube.exe!00007ff6cc049fd4()	Unknown
> > vkcube.exe!00007ff6cc046acf()	Unknown
> > vkcube.exe!00007ff6cc04f17e()	Unknown
> > kernel32.dll!00007fff933a7bd4()	Unknown
> > ntdll.dll!RtlUserThreadStart�()	Unknown
> 
> I'm still trying to figure out the other things though

if you build a debug version of the dll (graphics-hook64.dll) you will have symbols and you will be able to find the call which is on crash's way.

but naively looking a the callstack, it seems that you have other vulkan layer enabled, which may not interact properly with obs's layer (Bandicam Screen Recorder seems to be in the way : bdcamvk64.dll)

AluminumHaste commented:
I did have Bandicam installed, but I'm pretty sure I uninstalled it, I'll check tonight, thanks for pointing that out.

AluminumHaste commented:
Uninstalled Bandicam, still crashes but bdcamvk64 is no longer in the stack trace. 
Looks like I'm going to have to figure out how to compile this.

LittleMatth commented:
> Uninstalled Bandicam, still crashes but bdcamvk64 is no longer in the stack trace.
> Looks like I'm going to have to figure out how to compile this.
> EDIT: Okay, I started getting everything set up and it's way too complicated sorry.
> Can someone upload the pdb files instead?

do not hesitate to ask question, i can surely help, it looks like you are close to have everything setup.

Anths2016 commented:
Too bad that Retroarch crashes, if I change the game, and then it is no longer possible to restart it with SLOBS open.

Fenrirthviti commented:
> Too bad that Retroarch crashes, if I change the game, and then it is no longer possible to restart it with SLOBS open.

This project does not support or have anything to do with SLOBS. Please reach out to their support for more assistance with their fork.

jp9000 commented:
Did a major refactor of the source.  Pruned about 1600 lines from the main capture source file.  Cleaned everything up to the best of my ability.  Going to fix up the registry and backward compatibility stuff next.

samikhawaja commented:
Hi @jp9000,

I am thinking of pulling this and trying it on windows. What would you recommend, Is it too early or should I wait for the merge?

AluminumHaste commented:
Does it matter that the jsons target the 114 api? I have the 130 SDK installed.

samikhawaja commented:
I have tested this with Red Dead Redemption 2 and it seems to be working.

alinsavix commented:
With the recent changes, I'm still getting the "capture once per game run" behavior for Detroit: Become Human. Please let me know if there is any useful debugging information I can provide to assist in debugging this.

AluminumHaste commented:
I downloaded the artifact from here: https://dev.azure.com/obsjim/obsjim/_build/results?buildId=2371&view=logs&jobId=79941367-28af-51c2-dfb3-da6faf02333f&j=48cfd6ea-6219-5519-2b3d-ca7500f70a71

I still get the same problem with 64bit Vulkan games crashing on startup.

samikhawaja commented:
> With the recent changes, I'm still getting the "capture once per game run" behavior for Detroit: Become Human. Please let me know if there is any useful debugging information I can provide to assist in debugging this.

Can you please elaborate what do you mean by "capture once per game run". I am asking because I noticed a behavior where if during the game capture if I restart OBS, then the capture doesn't work and I get black screen. Is that what you mean here?

Leopard1907 commented:
You have to open OBS before booting the game. Because VLK layer needs to be loaded before game starts.

alinsavix commented:
> Can you please elaborate what do you mean by "capture once per game run". I am asking because I noticed a behavior where if during the game capture if I restart OBS, then the capture doesn't work and I get black screen. Is that what you mean here?

As described in https://github.com/obsproject/obs-studio/pull/2156#issuecomment-571321030

Mister36MN commented:
Is there any way to get a build of this with the changes that Jim made?

LangLangSAMA commented:
The vulkan capture is detected as cheating by BattlEye, is it possible to use anti-cheat compatibility hook for vulkan?

dodgepong commented:
That is because you are using an unsigned binary. The release binary will be signed and whitelisted by Battleye. You will be able to test this when the beta/RC is released in a few weeks.

clxss commented:
it's working ?

topeterk commented:
> That is because you are using an unsigned binary. The release binary will be signed and whitelisted by Battleye. You will be able to test this when the beta/RC is released in a few weeks.

The changes are made to an separate DLL only. Is BattleEye checking each loaded DLL? I mean, I was using the released OBS version and just copied the vulcan files to this installation. Would this be also detected by BattleEye as I am running the white listed EXE file?

dodgepong commented:
The DLL is a binary, and yes, BattlEye is checking them. In particular, this is the DLL handling the graphics hook. BattlEye doesn't only check EXE files, but also the libraries (DLLs) that are being loaded to interact with the game. The graphics hook is the most important of these, hence the reason it's getting flagged. When the DLL is signed (which can only be done by an official OBS build), then BattlEye will know that the DLL was built by a trusted source, and will allow it through.

LangLangSAMA commented:
> The DLL is a binary, and yes, BattlEye is checking them. In particular, this is the DLL handling the graphics hook. BattlEye doesn't only check EXE files, but also the libraries (DLLs) that are being loaded to interact with the game. The graphics hook is the most important of these, hence the reason it's getting flagged. When the DLL is signed (which can only be done by an official OBS build), then BattlEye will know that the DLL was built by a trusted source, and will allow it through.

Thank you for your explanation! It helps a lot!

EladBahar commented:
Hi,
does it working in R6 Vulkan?
thanks,

dodgepong commented:
As mentioned in the discussion above, yes it should work, provided you have a signed binary. A test build will be available in the coming weeks for you to test it yourself.

EladBahar commented:
OK thanks,
meanwhile i download the repository and compile it.
also i sign the graphics-hook64.dll  to test it, so it does work for dota2.
but in R6 vulkan it doesn't, 
i enable the out put debug and i can see the only "vkGetInstanceProcAddr .." hook is printed,
but no "vkGetDeviceProcAddr", and "QueuePresentKHR" never called

attach debug out put log
[obs-vulkan-logs-r6.LOG](https://github.com/obsproject/obs-studio/files/4166562/obs-vulkan-logs-r6.LOG)

Fenrirthviti commented:
You can't sign it yourself, you do not have our code sign certificate that is whitelisted by the anticheat solution that R6 uses. It's not just any signature, it's our signature. You have to wait until we provide a signed build.

jp9000 commented:
Rebased because cmake and issues with submodules were getting annoying.

jp9000 commented:
I also included code that now handles automatic installation of registry.  The installer and auto-updater should handle this automatically now, but just in case it's a portable zip build, it'll also attempt to install registry on startup.

AluminumHaste commented:
> 
> 
> > Uninstalled Bandicam, still crashes but bdcamvk64 is no longer in the stack trace.
> > Looks like I'm going to have to figure out how to compile this.
> > EDIT: Okay, I started getting everything set up and it's way too complicated sorry.
> > Can someone upload the pdb files instead?
> 
> do not hesitate to ask question, i can surely help, it looks like you are close to have everything setup.

Okay, finally got it all compiled (what a journey).

This is what I get in VS2019 when I debug run vkquake.exe
First on the stack is vulkan-1.dll
then:
graphics-hook64.dll!vk_shared_tex_supported(vk_inst_funcs * funcs, VkPhysicalDevice_T * phy_device, VkFormat format, unsigned int usage, VkExternalMemoryProperties * external_mem_props) Line 1032
	at c:\apps\obs\obs-studio\plugins\win-capture\graphics-hook\vulkan-capture.c(1032)

Line 1032 of that source file:
	VkResult result = funcs->GetPhysicalDeviceImageFormatProperties2KHR(
		phy_device, &info, &props);


Next down on the stack is:
graphics-hook64.dll!OBS_CreateDevice(VkPhysicalDevice_T * phy_device, const VkDeviceCreateInfo * cinfo, const VkAllocationCallbacks * ac, VkDevice_T * * p_device) Line 1298
	at c:\apps\obs\obs-studio\plugins\win-capture\graphics-hook\vulkan-capture.c(1298)

Line 1298:
	if (!vk_shared_tex_supported(ifuncs, phy_device, format, usage,
				     &data->external_mem_props)) {
		flog("texture sharing is not supported\n");
	} else {
		ret = VK_SUCCESS;
	}

Not sure if that helps.

jp9000 commented:
Are you saying that it's crashing on line 1032?

AluminumHaste commented:
I guess, I really don't know what I'm doing, just following instructions.

jp9000 commented:
Could you post a screenshot of when the debugging stops?

jp9000 commented:
And when I say when the debugging stops, I mean run it again, get it to reproduce the problem, then show me a screenshot of what it shows you when that happens at least.

AluminumHaste commented:
The game crashes immediately when trying to run the 64bit version of the app.

Screenshot
![VulkanQuake164bitcrash](https://user-images.githubusercontent.com/4510040/74117259-1c8da580-4b85-11ea-8bfd-d5349391d685.png)

AluminumHaste commented:
Here's what I get when I try to run the 64bit version of vkCube.exe from the Vulkan SDK.

![vkCube64bitcrash](https://user-images.githubusercontent.com/4510040/74117438-a63d7300-4b85-11ea-8ed7-33289b56564f.png)

AluminumHaste commented:
If I rename graphics-hook64.dll to graphics-hook64.dl_, the apps and games run:

![vkCube64bitworking](https://user-images.githubusercontent.com/4510040/74117599-34b1f480-4b86-11ea-980e-40040d18b1ae.png)

jpark37 commented on UI/win-update/updater/updater.cpp:
```diff
@@ -1071,6 +1071,74 @@ static bool UpdateVS2017Redists(json_t *root)
+	LSTATUS s;
+	HKEY key;
+
+	GetCurrentDirectoryW(_countof(path) - 1, path);
```
I don't think you need to subtract one for this API.

jp9000 commented:
Could you try again with this latest build?  Also please provide me with the debug output (in the view -> output window) of the vk cube sample app or an OBS log file.  Issue should be resolved though.  I'm curious to see which function you're unable to import, as I'm pretty sure that was the issue for you.  It should run now but it probably won't capture because you don't seem to support the required extension for whatever reason.

I'm just curious as to which extensions you don't seem to support, so if you could get me that debug output and/or OBS log file I'd appreciate it.

AluminumHaste commented:
Still not working.
![stillnotworking](https://user-images.githubusercontent.com/4510040/74287826-e2ddab80-4cf8-11ea-8b9a-0dec99c427d5.png)

AluminumHaste commented:
Dumb question, but the latest amd driver has Vulkan 1.2, which has had a bunch of extensions amalgamated into the core. This DLL was built against 1.1 I think, does that matter?

jp9000 commented:
Do you think you could come by discord and PM me?  Or email me?  I feel like I need to remote debug or teamview you or something to figure out what's going on there because none of us can reproduce your specific issue.  It seems like you're having a problem inside that function but I don't know why.  Being able to debug more in-depth or have more direct access would be ideal.

AluminumHaste commented:
I'll look into that for sure.
But for now, just some more info.
I just uninstalled all AMD software (except chipset drivers for motherboard), ran the AMD cleanup tool. Then I ran DDU and got rid of all Vulkan/AMD/Nvidia remnants.

Then I installed AMD driver 19.2.2 with the clean install option, and restarted the computer.
Still got the crash.
I then updated to the latest 20.2.1 driver, again with clean install option, same crash on 64bit vulkan applications. OBS itself runs fine.

I ran vulkaninfoSDK.exe (Uploaded to google drive: https://drive.google.com/open?id=1mh0b-vSSTJeQl6bGp4saDLZik34Ufm5Y) 
and vkvia.exe (https://drive.google.com/open?id=14iEjXR6NaOtD4XGO3zeIi6FSK2l4jT-_)

jp9000 commented:
Hm, possibly a driver issue?  Does it capture with 20.2.1?

AluminumHaste commented:
No, same issue. I'm in Discord now btw
Not really familiar with how it works

WizardCM commented:
If you run OBS in portable mode, then change the parent directory, the hook stops working, even though both paths are listed under Vulkan's Implicit Layers, and the new path of the JSON file is readable. This was tested with the Vulkan Cube demo. When I put it back into the original directory and restarted OBS, it started working again after a short time.

CoNoma commented:
I'm interested in testing this with Rainbow Six Siege now that it has a vulkan api that runs well. I have no idea how to set this up, but would be willing to take the time to figure it out, run it, see how it works and give you guys my logs. I run an amd cpu and amd gpu.

Mister36MN commented:
@CoNoma see messages above regarding R6. In short, won't be able to test until a digitally-signed version of OBS with Vulkan included is released, which will probably be a couple of weeks.

jpark37 commented on plugins/win-capture/graphics-hook/vulkan-capture.c:
```diff
@@ -0,0 +1,1598 @@
+			break;
+		}
+	}
+
```
Should handle mem_type_idx search failure.

AluminumHaste commented:
I'll try it again tonight.

AluminumHaste commented:
I'll try it again tonight.

jpark37 commented:
I can get vkcube working on my AMD card by switching VK_API_VERSION_1_0 to VK_API_VERSION_1_1. The issue is that the AMD ICD does not provide an internal callback for vkGetPhysicalDeviceImageFormatProperties2 for Vulkan 1.0 since this function was promoted to core in 1.1, and the Vulkan loader code is written to crash instead of gracefully failing.

OBS should probably scrape the apiVersion inside OBS_CreateInstance. If it's at least 1.1, then it can safely call the non-KHR function variant. If it's less than 1.1, then it should check the enumerated extensions to test if it's available.

FWIW, I've filed an issue with the Vulkan loader: https://github.com/KhronosGroup/Vulkan-Loader/issues/332

jp9000 commented:
Can we just change the API version from 1.0 -> 1.1 when calling the next vkCreateInstance?  (Only if it's 1.0 that is)  ...If it's just a minor API version number bump it should be backward-compatible right?

jpark37 commented:
Sounds scary; I'm not sure it's safe to bump the API version on behalf of the application. For example, could an extension disappear if it gets promoted to core, and now the application is calling a NULL function? Probably safer to co-exist happily within the application's tested environment rather than enforce our own.

jp9000 commented:
Just putting it up here for posterity, we already request 1.1 extensions, seems the code already took the 1.0 issue in to account:
```cpp
	/* -------------------------------------------------------- */
	/* enable instance extensions we need                       */

	struct ext_info req_ext[] = {
		{.name = VK_KHR_EXTERNAL_MEMORY_CAPABILITIES_EXTENSION_NAME},
		{.name = VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME},
	};
```

So that shouldn't be the issue.  There's something weird going on here.

jp9000 commented:
@AluminumHaste Sorry to ask this, but please test with games this time and not vk cube.

jp9000 commented:
It seems the entirety of our problems with the PR come down to stuff using vulkan 1.0.  1.1 seems to work fine.  If 1.0 causes us problems then I say we skip support for vulkan 1.0 entirely, because it's extremely unlikely that games are going to be using 1.0 at this point, especially newer games.

AluminumHaste commented:
https://www.pcgamingwiki.com/wiki/List_of_Vulkan_games

Doom 2016 and Ashes of the Singularity at version 1

jp9000 commented:
Somehow I doubt that doom 2016 has not updated their vulkan api version.

jp9000 commented:
Listen, if doom is still using 1.0 I'll figure something out.

Mister36MN commented:
Yeah I don't know if 2016 has updated to 1.1 either.

AluminumHaste commented:
> 
> 
> Somehow I doubt that doom 2016 has not updated their vulkan api version.

All the info I can find is that Doom was updated to use Vulkan version 1.0 back in 2016, and that was the last update to engine.
There seems to be a bunch of games using version 1.0, hopefully something can be figured out like you said.

jp9000 commented:
Just recording it here, Doom 2016 is definitely vulkan 1.1+.  Not Vulkan 1.0.

jpark37 commented on plugins/win-capture/graphics-hook/obs-vulkan32.json:
```diff
@@ -0,0 +1,82 @@
+            "DISABLE_VULKAN_OBS_CAPTURE": "1"
+        }
+    }
+}
```
Seems to be very different from obs-vulkan64.json. Should probably align the files.

jp9000 commented:
I haven't updated the 32bit json yet, I just took out the unused function hooks from the 64 json to test to see if it changed anything and it didn't

AluminumHaste commented:
I'm able to capture Doom 2016 Vulkan in borderless and fullscreen.
I do get an error sometimes when I start recording and I have to close and re-open OBS to continue recording.
I get a window popup that says:

"An unspecified error has occurred while recording."

In VS2017 I see some Catastrophic Errors in there when I stop recording and try to start again.

==== Recording Start ===============================================
Application provided invalid, non monotonically increasing dts to muxer in stream 1: 13012 >= -3

receive_audio: Error writing packet: Invalid argument
Output 'adv_ffmpeg_output': stopping
Output 'adv_ffmpeg_output': Total frames output: 5
Output 'adv_ffmpeg_output': Total drawn frames: 9
==== Recording Stop ================================================
The thread 0x359c has exited with code 0 (0x0).
'obs64.exe' (Win32): Loaded 'C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.18362.657_none_e6c5b579130e3898\comctl32.dll'. Cannot find or open the PDB file.
'obs64.exe' (Win32): Loaded 'C:\Windows\System32\WindowsCodecs.dll'. Cannot find or open the PDB file.
'obs64.exe' (Win32): Loaded 'C:\Windows\System32\thumbcache.dll'. Cannot find or open the PDB file.
The thread 0x240c has exited with code 0 (0x0).
The thread 0x2ea8 has exited with code 0 (0x0).
The thread 0x2d9c has exited with code 0 (0x0).
The thread 0x2da0 has exited with code 0 (0x0).
The thread 0x3054 has exited with code 0 (0x0).
The thread 0x2e9c has exited with code 0 (0x0).
The thread 0x31d8 has exited with code 0 (0x0).
The thread 0x3d40 has exited with code 0 (0x0).
The thread 0x335c has exited with code 0 (0x0).
The thread 0x3384 has exited with code 0 (0x0).
The thread 0x337c has exited with code 0 (0x0).
The thread 0x380c has exited with code 0 (0x0).
The thread 0x3bcc has exited with code 0 (0x0).
The thread 0x5a4 has exited with code 0 (0x0).
The thread 0x2e44 has exited with code 0 (0x0).
The thread 0x2d58 has exited with code 0 (0x0).
The thread 0x3378 has exited with code 0 (0x0).
The thread 0x33b0 has exited with code 0 (0x0).
The thread 0x3100 has exited with code 0 (0x0).
The thread 0x30d0 has exited with code 0 (0x0).
The thread 0x3018 has exited with code 0 (0x0).
The thread 0x311c has exited with code 0 (0x0).
The thread 0x3130 has exited with code 0 (0x0).
The thread 0x3110 has exited with code 0 (0x0).
The thread 0x302c has exited with code 0 (0x0).
The thread 0x3f54 has exited with code 0 (0x0).
The thread 0x2d00 has exited with code 0 (0x0).
The thread 0x35f8 has exited with code 0 (0x0).
The thread 0x35ac has exited with code 0 (0x0).
The thread 0x3478 has exited with code 0 (0x0).
The thread 0xba0 has exited with code 0 (0x0).
The thread 0x18ac has exited with code 0 (0x0).
The thread 0x3f4c has exited with code 0 (0x0).
final ratefactor: 8.70

'obs64.exe' (Win32): Loaded 'C:\Windows\System32\policymanager.dll'. Cannot find or open the PDB file.
'obs64.exe' (Win32): Loaded 'C:\Windows\System32\msvcp110_win.dll'. Cannot find or open the PDB file.
10 frames left in the queue on closing

'obs64.exe' (Win32): Loaded 'C:\Windows\System32\DriverStore\FileRepository\u0351505.inf_amd64_5b0d82e706ad63ce\B351435\amdxn64.dll'. Cannot find or open the PDB file.
The thread 0x36d4 has exited with code 0 (0x0).
onecore\windows\directx\database\helperlibrary\lib\perappusersettingsqueryimpl.cpp(121)\d3d9.dll!00007FFF7C670054: (caller: 00007FFF7C638EDC) ReturnHr(73) tid(25dc) 8000FFFF Catastrophic failure
onecore\windows\directx\database\helperlibrary\lib\perappusersettingsqueryimpl.cpp(98)\d3d9.dll!00007FFF7C66FF95: (caller: 00007FFF7C638E6A) ReturnHr(74) tid(25dc) 8000FFFF Catastrophic failure
onecore\windows\directx\database\helperlibrary\lib\directxdatabasehelper.cpp(999)\d3d9.dll!00007FFF7C66FF74: (caller: 00007FFF7C6371DB) ReturnHr(75) tid(25dc) 8000FFFF Catastrophic failure
'obs64.exe' (Win32): Unloaded 'C:\Windows\System32\DriverStore\FileRepository\u0351505.inf_amd64_5b0d82e706ad63ce\B351435\amdxn64.dll'

jp9000 commented:
I think it's unrelated to the capture.  The first part is related to an audio issue of some sort, and the second part doesn't seem to have anything to do with vulkan but instead D3D9 when the driver interface library loaded, but we don't use D3D9.  It's possible you had a video driver crash, usually it means some sort of system or driver instability but I don't suspect the problem lies in the capture itself at this point.  That's just my guess however.  We'll be having a lot of people test this soon.

jpark37 commented on plugins/win-capture/graphics-hook/obs-vulkan64.json:
```diff
@@ -0,0 +1,49 @@
+                "name": "VK_KHR_bind_memory2",
+                "spec_version": "1"
+            }
+        ],
```
If I understand correctly, these lists are for extensions implemented, not used. I would try removing all of this.

jpark37 commented on plugins/win-capture/graphics-hook/obs-vulkan64.json:
```diff
@@ -0,0 +1,49 @@
+        "name": "VK_LAYER_OBS_HOOK",
+        "type": "GLOBAL",
+        "library_path": ".\\graphics-hook64.dll",
+        "api_version": "1.1.114",
```
Should probably build against whatever the latest Vulkan SDK is, and bump this number.

jp9000 replied:
We don't actually link to vulkan-1.dll directly, it's all indirect through getprocaddress calls, so the version number here probably isn't super critical.

AluminumHaste commented:
I built with the latest changes, I can still capture Doom 2016 Vulkan, but when I try to game capture specific window and select vkQuake.exe from the list, it causes vkquake to crash.
OBS Log:

21:15:38.100: User added source 'Game Capture' (game_capture) to scene 'Scene'
21:15:43.845: [game-capture: 'Game Capture'] attempting to hook process: vkQuake.exe
21:15:43.886: [game-capture: 'Game Capture'] vk_shtex_init_d3d11_tex: OBS requesting VK_FORMAT_B8G8R8A8_UNORM texture format
21:15:43.888: [game-capture: 'Game Capture'] vk_shtex_init_vulkan_tex: BindImageMemory failed: VK_ERROR_OUT_OF_HOST_MEMORY
21:15:44.978: [game-capture: 'Game Capture'] capture window no longer exists, terminating capture
21:15:44.978: [game-capture: 'Game Capture'] capture stopped

I'm not sure which Vulkan version they're built against, (https://github.com/Novum/vkQuake), I compiled the 64bit binaries for vkquake yesterday. Game runs great otherwise.

jp9000 commented:
I tested with my AMD card and didn't have that issue.  I think you're using an older version of the hook (registry can be annoying like that) because that's the code we specifically fixed.

AluminumHaste commented:
Okay, I deleted the registry entries, now I can capture with the games crashing.
So we no longer need to JSON pointers in the registry?

![noJsonpointer](https://user-images.githubusercontent.com/4510040/74991856-334dbb00-5415-11ea-906b-7a96f0d2f96b.png)

jp9000 commented:
Without the game crashing you mean?  The registry entry was probably just pointing to an older library.  The registry is automatically recreated when the program starts up so it'll basically reset it in that case.

AluminumHaste commented:
Okay just finished recording a 30 minute session in vkQuake, no issues, working great.

jpark37 commented:
I have a lot of fixes for this PR, mostly to synchronization. Validation layer runs clean on vkcube, but who knows how actual games will behave: e98a1d4adee54da448fc7e43c2ebe0f2ed82c7b3

Will probably want to retest everything.

jpark37 commented:
Doom is poking holes in my fixes. I can take do another hardening pass tomorrow, so it's probably not worth testing this PR in the current state.

jpark37 commented:
Doom on AMD works again. I'm done for now.

AluminumHaste commented:
Will try and test some more vulkan apps tonight

jp9000 commented:
I'd probably recommend splitting off some of the highly-indented stuff in to a separate function, getting a bit squishy in there with the column limitations (sorry about that, should have used 120 way back when).  I can also do it later after you feel comfortable with your changes, although maybe you might have a better idea for the function name.

jpark37 commented:
Ok, I'll take care of it after work. I realized I need to handle swap chain size changes as well, and I have an idea to reorganize the data a bit too.

AluminumHaste commented:
Talos Principle works, captures, no issues that I've come across.
Vulkan Quake 1 still working well.
Doom 2016 works well also.

jpark37 commented:
Cleaned up the indentation, and added some untested code to handle theoretical swap chain growth.

I assume the present queue can handle copy operations. The Vulkan spec does not guarantee this, but the code to synchronize an extra copy queue properly involves Vulkan semaphores, and I don't want to learn them and muck up the code further unless we are forced to.

jp9000 commented:
Seems to work good.  Any other things remaining?

Mister36MN commented:
I just tried testing with R6 Vulkan. It doesn't give the error about the unsigned files, but if I have OBS open when starting R6, OBS just force closes with no crash report

AluminumHaste commented:
> 
> 
> I just tried testing with R6 Vulkan. It doesn't give the error about the unsigned files, but if I have OBS open when starting R6, OBS just force closes with no crash report

Is there anything in the log file?

Mister36MN commented:
> > I just tried testing with R6 Vulkan. It doesn't give the error about the unsigned files, but if I have OBS open when starting R6, OBS just force closes with no crash report
> 
> Is there anything in the log file?

Nope, no crash logs, nothing in normal logs. If I start OBS after running R6 Vulkan, it just fails to capture

09:50:23.544: [game-capture: 'Game Capture'] attempting to hook process: RainbowSix_Vulkan.exe
09:50:23.560: [game-capture: 'Game Capture'] capture stopped

AluminumHaste commented:
I think you can change the verbosity level of the logging.

jp9000 commented:
R6 has battleye so I would assume that's related somehow, probably not our fault.

dodgepong commented:
I believe @Mister36MN was using a signed capture hook binary, though. Are we positing that BattlEye isn't properly whitelisting this hook because it is Vulkan? Do we need to get in contact with them?

jp9000 commented:
I did get in contact with them, I think each individual game has to update.  Are there any other battleye games we could test with the newer hook dll?

I'm just saying that we don't get that with other games.  It doesn't even hook.

dodgepong commented:
Okay. I say that because the reason we gave for Vulkan not working on R6 in previous tests was because the people testing were not using signed hook binaries, but now we have a tester saying that even with a signed hook binary, it's still not working. Should we then get in contact with Ubisoft if we suspect it's something that's out of our control? Is there a way we can determine where the problem lies?

jpark37 commented:
The extension specs aren't exactly clear about how to fill out the pipeline barrier structs for shared images, but VK_QUEUE_FAMILY_EXTERNAL_KHR seems to be necessary, so i submitted a change to use it. That by itself causes the validation layer to complain in other ways, so I fixed things up the best I could.

Also, vkSetDeviceLoaderData exists for the dirty dispatch table fix-up on the command buffer, but it apparently can't be relied on to exist, so you'd need to keep the old code as a fallback path anyway. And who wants to execute a callback just to set a pointer? I decided not to bother.

AluminumHaste commented:
I must have been capturing OpenGL Quake2, as I tried last night with vkQuake2 (https://github.com/kondrak/vkQuake2) and capture fails to hook with a warning about usually caused by security software.
I already have OBS excluded and am only running the built in Windows 10 AV.
I didn't have a lot of time to test last night as the kids were not cooperating at bed time.

jp9000 commented:
I don't think this is an issue with the hook, for some reason vkquake2 isn't loading the layers.  The dll itself doesn't even load.  Other vulkan programs still work fine.  I don't know why only vkquake2 isn't loading layers.

Mister36MN commented:
> I did get in contact with them, I think each individual game has to update. Are there any other battleye games we could test with the newer hook dll?
> 
> I'm just saying that we don't get that with other games. It doesn't even hook.

I can try and see if there are any others that have Vulkan/BattleEye (or any AC I suppose) and see if I have them.

jp9000 commented:
We can verify if battleye isn't working on a program by just running it in a non-vulkan mode so we don't confuse whether it's a vulkan issue or a anti-cheat issue.

AluminumHaste commented:
> 
> 
> I don't think this is an issue with the hook, for some reason vkquake2 isn't loading the layers. The dll itself doesn't even load. Other vulkan programs still work fine. I don't know why only vkquake2 isn't loading layers.

I noticed in the logs that it shows:
21:33:09.054: load_graphics_offsets: Failed to start 'get-graphics-offsets32.exe'

I searched the build directory, that binary is not there. vkQuake2 is a 32bit application if that matters.

jp9000 commented:
Turns out it was our fault for that one, figured it out.  jpark was right about updating that API version in the json, I thought it worked differently.  vkQuake2 works now.

Also don't worry about that log message, you can only build either 32bit or 64bit, usually the builds are combined for the final OBS build.  Also if a game is 32bit, you need to compile the hook 32bit.  But vkquake2 is 64bit, not 32bit.  Don't run it in 32bit, there's little point.

AluminumHaste commented:
Ok got it.

I tried looking to compile the 32bit version just to see if that fixed it, but I don't have an x86 option in the build manager for any of the projects.
Sorry, I warned you I really don't know what I'm doing :P

LittleMatth commented:
> I don't think this is an issue with the hook, for some reason vkquake2 isn't loading the layers. The dll itself doesn't even load. Other vulkan programs still work fine. I don't know why only vkquake2 isn't loading layers.

Exactly, in the end it's up to the game developer.
Every vulkan layer come with a way to get disabled through environment variable.
So any game can scan the registry to discover available layers, blacklist some and whitlist some others.
Keep up the good work guys, what you are doing is amazing !

jp9000 committed:
win-capture/graphics-hook: Don't allow multiple hooks

Prevents multiple separate hook DLLs from being loaded in to a target at
once.  This is done just in the off-chance that someone might add
another hook to the Vulkan layer registry.

jp9000 committed:
cmake: Add FindVulkan module

jp9000 committed:
win-capture: Use full DLL path for inject helper

The inject helper should be able to specify the full path rather than
assume the path of the hook DLL.  This change allows us to modify the
hook's location.  This needs to be done because the hook needs to be
relocated to ProgramData to prevent the possibility of multiple Vulkan
capture hooks.

LittleMatth committed:
win-capture: Add Vulkan capture

Co-authored-by: jp9000 <obs.jim@gmail.com>
Co-authored-by: jpark37 <jpark37@users.noreply.github.com>

jpark37 commented:
I have a couple ideas for dealing with versioning.

Separate paths for obs-studio-hook versions:
- Copy to C:\ProgramData\obs-studio-hook\<version>\ instead of C:\ProgramData\obs-studio-hook\
- Pro: Much less likely for copy to fail because an application or game has the hook loaded.
- Pro: No risk of old hook files smashing newer hook files.
- Con: Hook DLLs will probably never be cleaned up.

Alternatively, only allow one active hook (RenderDoc style?):
- When Game Capture initializes, scan HKLM and HKCU for OBS entries. If there is only one, and it's for the current OBS installation, then it's all good. Otherwise, make button visible in Game Capture dialog to "Register Vulkan hook for current location" or something.
- If button is pressed, attempt to remove existing registry entries. Use separate ShellExecuteEx/runas process if HKLM entry exists and HKLM removal fails. If all entries cannot be removed, alert user of failure.
- If existing entries are removed, and user is admin, attempt to register to HKLM. Use separate ShellExecuteEx/runas if OBS process is not elevated.
- If user is not admin, inform user that registration can not be applied system-wide, and if just user-wide registration is fine.
- Whether successful or not, inform user and update registration button visibility.
- Pro: No need to copy to C:\ProgramData
- Pro: No risk of hook files smashing other hook files.
- Pro: No orphaned hook files.
- Pro: User has control over which OBS installation is Vulkan master.
- Con: Good amount of code needs to be written.

mufunyo commented:
I'm still in favour of just having one unversioned hook, and just replacing it every time. If you're concerned with file sharing violations, just rename existing hook DLLs and mark them for deletion on next reboot.

Fenrirthviti commented:
I am as well, but I believe the concern is anti-cheat. Having multiple hook versions allows us to fall back on an older hook that is whitelisted already when games are slow to updated anti-cheat signatures. Personally, I am torn on this because I don't feel it's our responsibility to ensure that OBS works with the game, especially when the solution is turning out to be so complicated/fragile. But at the same time, if a simple solution can be provided, it probably can't hurt.

jp9000 committed:
win-capture: Check hook version before capture init

Checks the hook version to ensure compatibility with hook DLL.  It's
unlikely it'll ever be necessary to increment the hook version, but this
is just a precautionary thing that allows a hook DLL to make sure it's
rejected by an older OBS version if needed.  Again however, very
unlikely that the major version will ever be incremented.

MartB commented:
Rainbow six displays this at startup:
12:45:55: [INFO] Blocked loading of file: "C:\ProgramData\obs-studio-hook\graphics-hook64.dll".

MartB commented:
12:55:04: [INFO] Blocked loading of file: "C:\Program Files\OBS Studio (64bit)\data\obs-plugins\win-capture\graphics-hook64.dll".

Is there any way to get a signed hook dll?

jp9000 commented:
If we need separated files I'll probably just use a new filenames, like graphics-hook64-v2.dll or something.  I don't think folders are necessary.  I think this iteration should be sufficient until then.

jp9000 commented:
@martB -- Ubisoft will have to unblock it when they're ready to do so as far as I'm aware.  I am in contact with them so I'll ask them about it when the time comes.

mufunyo commented:
I strongly feel that anti-cheat software should be whitelisting OBS hooks based on our signature, not on a file-by-file basis. Because of the nature of the Vulkan layer registry, having multiple files, versioned or not, and having to delete duplicate/spurious entries is an option that is just going to explode in our faces later down the line.

The hook itself does very little and should not have to be updated often, and if an update breaks anti-cheat software, users can take it up with the game vendor or the anti-cheat vendor to get the update whitelisted. I don't think we should bend over backwards to avoid breaking anti-cheat compatibility when it's likely the solution will break Vulkan game capture in other horrible ways in the future.

jp9000 commented:
Just for the record I doubt it'll ever come to that.  I just meant in the absolute need we can do that.  I doubt we're going to be changing the hook significantly apart from bug fixes (famous last words)

AluminumHaste commented:
Just want to say that this has been a really fun ride and I learned a lot. I'll be happy to help test in the future! Thanks for you patience jp9000

AluminumHaste commented:
> 
> 
> I strongly feel that anti-cheat software should be whitelisting OBS hooks based on our signature, not on a file-by-file basis. 

Wouldn't that make it too easy for a cheat developer to download OBS source, modify the hook for their purposes and then it would be whitelisted and not caught?

mufunyo commented:
> Wouldn't that make it too easy for a cheat developer to download OBS source, modify the hook for their purposes and then it would be whitelisted and not caught?

The OBS project signature is issued to "Hugh Bailey", and only Jim can sign binaries with that signature. Any modified binaries would be unsigned and immediately caught by anti-cheat.

AluminumHaste commented:
Got it.

dstensnes commented:
Just out of curiosity: For websites, browsers only have to include root CAs from each SSL certificate vendor (e.g. lets-encrypt), and browsers will then trust all certificates issued by them (certificate chain system). Is it the same with these anti-cheat systems and signed files, or do you have to get a valid signature for each file, no matter how small the diff?

mufunyo commented:
I have been informed that the whitelisting issues we encountered were due to Jim's old certificate being expired and having to obtain a new certificate and getting it whitelisted. The new certificate is valid until 2021 so we shouldn't encounter any problems until then. To answer your question, the anti-cheat whitelisting is per certificate, so any binary signed with the current certificate will be whitelisted.

DePhoegon commented:
> Couldn't resist: VM installed with Win10, downloaded all sources, IDEs and more dependencies, patched commits of this issue and compiled it.
> After running the batch, starting OBS Studio and then starting the game, I got a video from the game capture of "_The Surge 2 (Vulkan, 64Bit)_". This is great!
> 
> If someone else wants to try this, I attached my built **Win10x64** binaries of the folder "_data\obs-plugins\win-capture_" which should be enough to get it running. Remember to run the batch to activate the hook (and start OBS before the game).
> [OBSStudio24.0.3 VulkanGameGapture-caff574 Win10x64.zip](https://github.com/obsproject/obs-studio/files/3870902/OBSStudio24.0.3.VulkanGameGapture-caff574.Win10x64.zip)

The bat file fails because you need to put " " around the value after the reg commands. 

works perfectly for Doom Eternal  & Streamlabs OBS (though it's default folder you need is in "C:\Program Files\Streamlabs OBS\resources\app.asar.unpacked\node_modules\obs-studio-node\data\obs-plugins\win-capture "
